vocabularies/DirectEdit.xml

```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="utf-8"?>
+<edmx:Edmx xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" Version="4.01">
+  <edmx:Reference Uri="https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Core.V1.xml">
+    <edmx:Include Alias="Core" Namespace="Org.OData.Core.V1" />
+  </edmx:Reference>
+  <edmx:Reference Uri="https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Validation.V1.xml">
+    <edmx:Include Alias="Validation" Namespace="Org.OData.Validation.V1" />
+  </edmx:Reference>
+  <edmx:Reference Uri="https://sap.github.io/odata-vocabularies/vocabularies/Common.xml">
+    <edmx:Include Namespace="com.sap.vocabularies.Common.v1" Alias="Common" />
+  </edmx:Reference>
+
+  <edmx:DataServices>
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="com.sap.vocabularies.DirectEdit.v1" Alias="DirectEdit">
+      <Annotation Term="Core.Description">
+        <String>Terms for Direct-Edit User Interfaces</String>
+      </Annotation>
+      <Annotation Term="Core.Description" Qualifier="Published">
+        <String>2020-07-10 © Copyright 2020 SAP SE. All rights reserved</String>
+      </Annotation>
+      <Annotation Term="Core.Links">
+        <Collection>
+          <Record>
+            <PropertyValue Property="rel" String="latest-version" />
+            <PropertyValue Property="href" String="https://sap.github.io/odata-vocabularies/vocabularies/DirectEdit.xml" />
+          </Record>
+          <Record>
+            <PropertyValue Property="rel" String="alternate" />
+            <PropertyValue Property="href" String="https://sap.github.io/odata-vocabularies/vocabularies/DirectEdit.json" />
+          </Record>
+          <Record>
+            <PropertyValue Property="rel" String="describedby" />
+            <PropertyValue Property="href" String="https://github.com/sap/odata-vocabularies/blob/master/vocabularies/DirectEdit.md" />
+          </Record>
+        </Collection>
+      </Annotation>
+
+      <Term Name="SideEffects" Type="DirectEdit.SideEffectsType" Nullable="false" AppliesTo="EntitySet">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="Determine side effects of client-side data modification" />
+      </Term>
+
+      <ComplexType Name="SideEffectsType">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.LongDescription">
+          <String>A direct-edit side-effect consists of:
+
+- a list of possible triggers, _preliminary_ syntax:
+  - `Path:` followed by a path to the property whose change triggers the side-effects determination 
+  - `Action:` followed by the name of the action that is called for the side-effects determination
+
+- an action bound to the entity set with the following non-binding parameters:
```

Comment:
Isn't it sufficient if every entity set has _one_ action `SideEffects` whose single parameter `Data` and return type both have the given entity type (or something structure-compatible)? For example,
```
POST ~/TimeSheet/SideEffects
Content-Type: application/json

{"Data": {"TimeType": "sickness"}}
```
would return
```json
{"@odata.context": "$metadata#TimeSheet",
 "value": {"TimeType": "sickness", "UploadEnabled": true}}
```
Then `DirectEdit.SideEffects` would be unnecessary and the long list of side effects need not be passed to the client, only the server must implement it. (Or, if the client wants to avoid unnecessary requests, it only needs the list of triggers, not the calculation actions. But I think that in the end, _every_ change on the client will trigger some side effect.)

Reply:
The list of actions is not completely necessary but it gives more flexibility. 
The triggers are important. Every field change with a side effect has a performance impact so we want to avoid unnecessary round trips.
We do need to know which field triggered the change because this is not possible to determine just from the backend state (e.g. if you have a side effect when changing the position and another one when changing the division, you do not want to recalculate the position change after a subsequent division change).

Reply:
> We do need to know which field triggered the change

But that could be achieved by having _two_ parameters `Data` and `BeforeData`, which would not only show which property/ies has/ve changed, but also the previous value(s), which could also be important.

Reply:
At least for our current non-OData based implementation we do not really need to know the complete before image but simply send the information which field changed.
The before image would also need more work in the client and would roughly double the payload for most cases.
I do not want to discount other use cases, though. Do you see a way of adding additional variants later?

Reply:
I want to keep it simple, not just for now but also long-term. Based on my experience with `Common.SideEffects`, I am afraid that this annotation might grow into a monster, therefore I would like to avoid it altogether.

Better to keep the complete knowledge on the server and have the client invoke an action with `Data` and `BeforeData` after every change. If your current use case doesn't care about the before-value, you can even set it equal to the after-value, the mere presence of `BeforeData/SomeProperty` would then indicate that this property triggered the action call. And if `BeforeData` contains only the `SomeProperty`, its effect on the payload is negligible.

Reply:
My proposal was to send `BeforeData` = before-image and `Data` = after-image. The `BeforeData` could be restricted to the properties that have actually changed, thus implying the trigger.